doc/neps/nep-00XX-array-function-protocol.rst

```diff
@@ -0,0 +1,538 @@
+==================================================
+NEP: Dispatch Mechanism for NumPy's high level API
+==================================================
+
+:Author: Stephan Hoyer <shoyer@google.com>
+:Author: Matthew Rocklin <mrocklin@gmail.com>
+:Status: Draft
+:Type: Standards Track
+:Created: 2018-05-29
+
+Abstact
+-------
+
+We propose a protocol to allow arguments of numpy functions to define
+how that function operates on them. This allows other libraries that
+implement NumPy's high level API to reuse Numpy functions. This allows
+libraries that extend NumPy's high level API to apply to more NumPy-like
+libraries.
+
+Detailed description
+--------------------
+
+Numpy's high level ndarray API has been implemented several times
+outside of NumPy itself for different architectures, such as for GPU
+arrays (CuPy), Sparse arrays (scipy.sparse, pydata/sparse) and parallel
+arrays (Dask array) as well as various Numpy-like implementations in the
+deep learning frameworks, like TensorFlow and PyTorch.
+
+Similarly there are several projects that build on top of the Numpy API
+for labeled and indexed arrays (XArray), automatic differentation
+(Autograd, Tangent), higher order array factorizations (TensorLy), etc.
+that add additional functionality on top of the Numpy API.
+
+We would like to be able to use these libraries together, for example we
```

Comment:
Add:
```
Finally, some ``ndarray`` subclasses add new behaviour (e.g., MaskedArray and astropy's Quantity),
which gives different meaning to functions. Indeed, MaskedArray reimplements some of the core
numpy functions, and Quantity has long-standing issues about compatibility with stacking functions.
```

Reply:
I'm not sure how best to weave this text into the NEP.  To me this seems like orthogonal functionality.

Reply:
My text may have been confusing, but I think `__array_function__` is definitely *very* relevant for subclasses. 

For instance, currently `ma.extras` is full of re-definitions of functions like `stack` - i.e., it follows the make-your-own-namespace "solution" that this NEP would like to address. This could be completely replaced by the `__array_function__` proposed here!

Similarly, for `Quantity` we are currently relying on overrides or hope that a function uses ufuncs and does not do `asarray` - and we have no good solution for functions like `concatenate` (not being willing to provide our own namespace...). With `__array_function__`, this would be solved. For instance, for `concatenate` we can ensure that all inputs are converted to the same unit before doing the actual concatenation.

In other words, I probably should have written what looks much like this NEP for `MaskedArray` and `Quantity` alone - hence the request to mention them!

Aside: For both the above, I see possible implementations very much in terms of "1. prepare input arrays; 2. call original function on those; 3. set some properties on outputs" - step (2) is most logically implemented with a `super()` call, which is why I strongly suggest there be an `ndarray.__array_function__`.